R/pkg.R

```diff
@@ -257,13 +264,30 @@ getPackageRecords <- function(pkgNames,
   # Remove any null records
   allRecords <- dropNull(allRecords)
 
+  if (.recursion.level == 1) {
+    .nnn <- length(allRecords)
+    .iii <- 1
+  }
+
+
   # Now get recursive package dependencies if necessary
   if (recursive) {
     allRecords <- lapply(allRecords, function(record) {
+      if (.recursion.level == 1 && verbose) {
+        logger(sprintf("- getting dependency %3i of %3i %s", .iii, .nnn, record$name))
+        .iii <<- .iii + 1
+      }
       deps <- getPackageDependencies(pkgs = record$name,
                                      lib.loc = lib.loc,
                                      available.packages = available)
-      if (!is.null(deps)) {
+
+      new  <- setdiff(deps, ls(envir = .visited.packages))
+      for (pkg in new) {
```

Comment:
We probably should update `.visited.package` based on the incoming set of packages (`pkgNames`). The approach here can still visit a package repeatedly (at most twice).

Filtering previously visited packages (the `setdiff` call) can still remain here (but could also shift to the start of the function).